[docs] Add a reference guide custom build configs (#21767)
expo/expo: docs/pages/build-reference/custom-build-config.mdx

---
title: Custom EAS Build configs
description: Learn how to create and use custom workflows with EAS Build by creating custom build configs.
hideFromSearch: true
---

import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';
import ImageSpotlight from '~/components/plugins/ImageSpotlight';
import { BoxLink } from '~/ui/components/BoxLink';
import { GithubIcon } from '@expo/styleguide-icons';

> Custom workflows allow running tests on the cloud, adding instructions for native platforms, resigning a build to share with your team or multiple test devices, and so on. With EAS Build, you can create build configs to customize the process for adding your workflows. It helps saving time especially when working teams.

Author: My bad and you are right. It should be "when working in teams."

> Custom workflows allow running tests on the cloud, adding instructions for native platforms, resigning a build to share with your team or multiple test devices, and so on. With EAS Build, you can create build configs to customize the process for adding your workflows. It helps saving time especially when working teams.

Author: ```suggestion
Custom workflows allow running tests on the cloud, adding instructions for native platforms, resigning a build to share with your team or multiple test devices, and so on. With EAS Build, you can create build configs to customize the process for adding your workflows. It helps saving time especially when working in teams.
```

> This guide shows how to create and use a custom build config with EAS. It follows a scenario where you create a custom config to run tests on EAS Build. You can also use the defined steps in other scenarios.

Reviewer: ```suggestion
This guide shows how to create and use a custom build config with EAS. It follows a scenario where you create a custom workflow to run tests on EAS Build. You can also use the defined steps in other scenarios.
```

</Step>

<Step label="2">

## Create a workflow

You must add a build config file in your project to create a workflow. Create a directory path **.eas/build** at your project's root. Both path and name of both directories are important for EAS Build to identify that a project contains a custom build config.

Inside it, you can create a new file called **test.yml**. This file contains the workflow steps that you want to run. The file's name is unimportant; you can name it whatever you want. The only requirement is that the file extension uses **.yml**.

This file will contain steps to run tests on the EAS Build service. It'll install all the required dependencies from the **package.json** file of your project and execute the `npm test` script.

Add the following workflow steps in the file:

```yaml .eas/build/test.yml
build:
  name: Run tests
  steps:
    - run:
        name: Install dependencies
        command: npm install
    - run:
        name: Run tests
        command: |
          echo "Running tests..."
          npm test
```

</Step>

<Step label="3">

## Add `config` property in eas.json

> To use the custom build config, you must add the `config` property in **eas.json**. It contains the build config file's name as the value you want to run on the EAS.

Reviewer: ```suggestion
To use the custom build config, you must add the `config` property in **eas.json**. It contains the build config filename as the value you want to run on the EAS.
```

      "android": {
        "config": "test.yml"
      }
    },
}
```

</Step>

<Step label="4">

## Run build to test the workflow

To test the workflow, run the following command:

<Terminal cmd={['$ eas build -p android -e test']} />

After the build is complete, you can verify that the `npm test` script was executed by checking the logs on the **build details** page.

<ImageSpotlight
  alt="Example of a build config running tests on EAS Build."
  src="/static/images/eas-build/custom-build.png"
/>

</Step>

## Summary

- Build configs are customizable by adding steps you want to run on the EAS service as an extension to your existing regular builds.
- When the **eas.json** file contains the `config` property, credentials such as the keystore setup for Android and distribution certificate for iOS are skipped. This is not implemented yet because both the keystore and distribution certificate are unavailable on the server side.
- If an existing project already has credentials (keystore and distribution certificate), they remain unchanged.
- If your regular builds have [EAS secrets](/build-reference/variables/#adding-secrets-with-eas-cli), they are also available for custom builds.

> Also, check out the **example repository** for a more detailed example:

Reviewer: ```suggestion
Also, check out the **example repository** for more detailed examples:
```
